compiler/rustc_query_impl/src/lib.rs

```diff
@@ -36,13 +42,176 @@ use rustc_query_system::query::*;
 pub use rustc_query_system::query::{deadlock, QueryContext};
 
 pub use rustc_query_system::query::QueryConfig;
+use rustc_query_system::HandleCycleError;
 
 mod on_disk_cache;
 pub use on_disk_cache::OnDiskCache;
 
 mod profiling_support;
 pub use self::profiling_support::alloc_self_profile_query_strings;
 
+struct DynamicQuery<C: QueryCache + RemapQueryCache> {
```

Comment:
Could you put the introduction of this struct and related changes in a separate commit from the type-erasure change?

Reply:
Given the feedback by @BoxyUwU I'll split the `rustc_query_system` and virtual dispatch out to separate PRs.